activerecord/lib/active_record/core.rb

```diff
@@ -680,21 +683,14 @@ def connection_handler
       self.class.connection_handler
     end
 
-    # Returns the contents of the record as a nicely formatted string.
+    # Returns the attributes specified by #attributes_for_inspect as a nicely formatted string.
     def inspect
-      # We check defined?(@attributes) not to issue warnings if the object is
-      # allocated but not initialized.
-      inspection = if defined?(@attributes) && @attributes
-        attribute_names.filter_map do |name|
-          if _has_attribute?(name)
-            "#{name}: #{attribute_for_inspect(name)}"
-          end
-        end.join(", ")
-      else
-        "not initialized"
-      end
+      inspect_with_attributes(attributes_to_inspect)
+    end
 
-      "#<#{self.class} #{inspection}>"
+    # Returns the full contents of the record as a nicely formatted string.
+    def full_inspect
```

Comment:
What if, instead of a separate method, we had an `:all` option? 🤔 e.g. `Post.attributes_to_inspect = :all`. That would make it configurable without needing a separate config.

Reply:
I think there's a good argument for all of the above.

If a user wants the default behaviour of `.inspect` to be `:all` they should be able to configure it as such without listing all the attributes in the model `attributes_to_inspect = ` as those can change.

A user should also be able to decide to opt into a detailed inspect, without using it as the default.

In this example case we want to make the default to not be harmful, and not accidentally load all attributes deeply for all posts when logging a list:
```ruby
Rails.logger.info("Found these posts: #{ posts.map(&:inspect) }")
```

However, in some examples we do want to really see what's happening, even if it is an expensive operation:
```ruby
Rails.logger.error("This post caused a render error: #{ post.full_inspect }")
```

And finally, we do not want to take away tools from someone who knows what they want. This example model is simple and when we inspect it we really want to see all fields at all times, even if a field gets added or renamed:
```ruby
class Tag < ApplicationRecord
  validates :value, :handle, presence: :true
  attributes_to_inspect = :all
end
```

Reply:
So just to make sure I understand the suggestion, you want to have `:all` in the `attributes_to_inspect` array. And then inspect if `:all` is an element of the array, and if it is, then do a full inspection with all the attributes?

Personally, I dislike using the same array for the list of attributes and for a boolean option like `all`. One issue that comes to mind is what if the model has an attribute called `all`? I guess in `inspect` we can check if the `attributes_to_inspect` array has only one member, and if that one member is `all` then do the full inspection. But now the behaviour is getting convoluted IMO.

What do you guys think about my suggestion that people can just do `alias_method :inspect, :full_inspect` if they always want to do the full inspection by default. Adding it to `ApplicationRecord` would make the upgrade easy.

Reply:
`:all` would be passed as the option, rather than an array. ie. `Post.attributes_to_inspect = :all`. (`[:all]` would mean "an attribute called all")

@kmcphillips makes a great point about adding support for both `:all` and a `full_inspect` method though -- being able to specify a full inspect as necessary without needing to configure a class attr on the model makes sense for sure.

So maybe:

#### Global config

Use `config.active_record.attributes_to_inspect` to configure the default set of attributes to inspect across all your models. Will default to `[:id]`. Use `config.active_record.attributes_to_inspect = :all` to show all attributes by default.

####  Model-level config

Use `attributes_to_inspect` to configure the default set of attributes to inspect for a given model. e.g the example Kevin provided:
```ruby
class Tag < ApplicationRecord
  validates :value, :handle, presence: :true
  attributes_to_inspect = :all
end
```
#### `#full_inspect`

Use `#full_inspect` to do a full inspection, no matter the global / model config for `attributes_to_inspect`.

Reply:
IMO, a much more flexible approach would be to make `inspect` accept an optional argument:
```ruby
def inspect(attributes_to_list = [:id])
  if attributes_to_list == :all
    # do full inspection
  else
    # list only the listed attributes
  end
end
```

That way, callers that need the full behaviour will be able to get it by calling `post.inspect(:all)`. In this way the decision can be made per call site, instead of per model. For example, as well as being able to make the  `post.inspect(:all)` call, you could also do a `post.inspect([:id, :title])` for the same `Post` instance, if you wanted to.

Reply:
@paracycle The call site is often in gem / stdlib code, not application code. So I believe the developer would be locked into just inspecting with in `id` in those cases. The original issue came up in a case where we call `to_s` on a hash containing AR records, which calls inspect on the records. Also, it doesn't help make the behaviour easily opt out-able, since the developer would have to find all the call sites and add the `:all` parameter - and in cases like the above it wouldn't be possible to opt out.

Tho maybe it would be nice to have the ability to pick different attributes at specific call sites, and use the model's `attributes_to_inspect` as the default. It's not exactly the problem I am trying to solve, but I could see it being useful.

The name `attributes_to_list` is good thank you, I am going to use that.

@adrianna-chang-shopify Thanks for clarifying the `:all` option vs array distinction. After chewing it over I like the idea so I will hack on it sometime today or tomorrow 👍

Reply:
Isn’t the problem here that calling Hash#to_s calls inspect on its values instead of to_s?

Reply:
`Hash#to_s` is just an alias for `inspect` - so it doesn't seem unexpected to me it is calling `inspect` under the hood. Tho yeah, the issue would be fixed if `Hash#to_s` called `to_s` (tho personally I think the outcomeShop would be worse, since `to_s` on AR record doesn't return much useful information imo and the output from inspect is better).